pkg/settlement/swap/erc20/erc20.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package erc20
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/settlement/swap/transaction"
+	"github.com/ethersphere/sw3-bindings/v3/simpleswapfactory"
+)
+
+var (
+	erc20ABI     = transaction.ParseABIUnchecked(simpleswapfactory.ERC20ABI)
+	errDecodeABI = errors.New("could not decode abi data")
+)
+
+type Service interface {
+	BalanceOf(ctx context.Context, address common.Address) (*big.Int, error)
+	Transfer(ctx context.Context, address common.Address, value *big.Int) (common.Hash, error)
+}
+
+type erc20Service struct {
+	backend            transaction.Backend
+	transactionService transaction.Service
+	address            common.Address
+}
+
+func New(backend transaction.Backend, transactionService transaction.Service, address common.Address) Service {
+	return &erc20Service{
+		backend:            backend,
+		transactionService: transactionService,
+		address:            address,
+	}
+}
+
+func (c *erc20Service) BalanceOf(ctx context.Context, address common.Address) (*big.Int, error) {
+	callData, err := erc20ABI.Pack("balanceOf", address)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := c.transactionService.Call(ctx, &transaction.TxRequest{
+		To:   &c.address,
+		Data: callData,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := erc20ABI.Unpack("balanceOf", output)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(results) != 1 {
+		return nil, errDecodeABI
+	}
+
+	balance, ok := abi.ConvertType(results[0], new(big.Int)).(*big.Int)
+	if !ok || balance == nil {
+		return nil, errDecodeABI
+	}
+	return balance, nil
+}
+
+func (c *erc20Service) Transfer(ctx context.Context, address common.Address, value *big.Int) (common.Hash, error) {
+	callData, err := erc20ABI.Pack("transfer", address, value)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	request := &transaction.TxRequest{
+		To:       &c.address,
+		Data:     callData,
+		GasPrice: nil,
```

Comment:
if the `GasPrice` is `nil` and `GasLimit` is `0` then what happens to the transaction once it gets sent? Not really clear what is the expected behavior here?

Reply:
those are the values for the transaction request, not the transaction itself. gas price nil means the transaction service will query the ethereum node for a gas price, gas limit 0 means it will do a gas estimation.

Reply:
cool. maybe good to document this inline?